File: 802-Find-Eventual-Safe-States.py
class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        # adj=[[] for i in range(len(graph))]
        # for i in range(len(graph)):
        #     ad[i].append(graph[i])
        visited=[0]*len(graph)
        path=[0]*len(graph)
        li=[]
        def dfs(i):
            visited[i]=1
            path[i]=1
            for node in graph[i]:
                if visited[node]==0:
                    if dfs(node) is False:
                        return False
                elif path[node]==1:
                    return False
            path[i]=0
            li.append(i)
            return
        for i in range(len(graph)):
            if visited[i]==0:
                dfs(i)
        li.sort()
        return li

File: 802-Find-Eventual-Safe-States.cpp
class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        vector<vector<int>> reversed_graph(graph.size());
        vector<int> indegree(graph.size(),0);
        queue<int> q;
        vector<int> op;
        for(int i=0;i<graph.size();i++){
            for(int j=0;j<graph[i].size();j++){
                reversed_graph[graph[i][j]].push_back(i);
                indegree[i]++;
            }

        }
        for(int i=0;i<indegree.size();i++){
            if(indegree[i]==0){
                q.push(i);
            }
        }
        while(!q.empty()){
            int val=q.front();
            
            op.push_back(val);
            q.pop();
            for(int i=0;i<reversed_graph[val].size();i++){
                indegree[reversed_graph[val][i]]--;
                if(indegree[reversed_graph[val][i]]==0){
                    q.push(reversed_graph[val][i]);
                }
            }
        }
        sort(op.begin(),op.end());
        return op;


        
    }
};